hudi-common/src/main/java/org/apache/hudi/common/util/TablePathUtils.java

```diff
@@ -92,12 +92,21 @@ private static boolean isInsideTableMetadataFolder(String path) {
         HoodiePartitionMetadata metadata = new HoodiePartitionMetadata(fs, partitionPath);
         metadata.readFromFS();
         return Option.of(getNthParent(partitionPath, metadata.getPartitionDepth()));
+      } else {
+        // Simply traverse directory structure until found .hoodie folder
+        Path current = partitionPath;
+        while (current != null) {
+          if (hasTableMetadataFolder(fs, current)) {
```

Comment:
One caveat is that this may incur more than one `fs.exists()` calls.  Is this only used for initialization (which is fine), e.g., getting table path from config, and not for core read/write logic per data file?

Reply:
Correct. This is only useful in a discovery phase.